// write-excel-file/modules/xlsx/files/sheet.xml/helpers/getCellStyleProperties.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import addProperties from '../../../helpers/features/addProperties.js';
import hasAlignment from '../../../helpers/hasAlignment.js';
import hasBorder from '../../../helpers/hasBorder.js';
import hasFill from '../../../helpers/hasFill.js';
import hasFont from '../../../helpers/hasFont.js';
export default function getCellStyleProperties(cell, features) {
  var align = cell.align,
    alignVertical = cell.alignVertical,
    textRotation = cell.textRotation,
    indent = cell.indent,
    wrap = cell.wrap,
    fontFamily = cell.fontFamily,
    fontSize = cell.fontSize,
    fontWeight = cell.fontWeight,
    fontStyle = cell.fontStyle,
    textDecoration = cell.textDecoration,
    textColor = cell.textColor,
    backgroundColor = cell.backgroundColor,
    fillPatternStyle = cell.fillPatternStyle,
    fillPatternColor = cell.fillPatternColor,
    borderColor = cell.borderColor,
    borderStyle = cell.borderStyle,
    leftBorderColor = cell.leftBorderColor,
    leftBorderStyle = cell.leftBorderStyle,
    rightBorderColor = cell.rightBorderColor,
    rightBorderStyle = cell.rightBorderStyle,
    topBorderColor = cell.topBorderColor,
    topBorderStyle = cell.topBorderStyle,
    bottomBorderColor = cell.bottomBorderColor,
    bottomBorderStyle = cell.bottomBorderStyle;

  // Add any custom cell style properties that're used by plugins.
  var additionalProperties = {};
  addProperties(additionalProperties, cell, features, 'style');
  if (hasAlignment({
    align: align,
    alignVertical: alignVertical,
    textRotation: textRotation,
    indent: indent,
    wrap: wrap
  }) || hasFont({
    fontFamily: fontFamily,
    fontSize: fontSize,
    fontWeight: fontWeight,
    fontStyle: fontStyle,
    textDecoration: textDecoration,
    textColor: textColor
  }) || hasFill({
    backgroundColor: backgroundColor,
    fillPatternStyle: fillPatternStyle,
    fillPatternColor: fillPatternColor
  }) || hasBorder({
    borderColor: borderColor,
    borderStyle: borderStyle,
    leftBorderColor: leftBorderColor,
    leftBorderStyle: leftBorderStyle,
    rightBorderColor: rightBorderColor,
    rightBorderStyle: rightBorderStyle,
    topBorderColor: topBorderColor,
    topBorderStyle: topBorderStyle,
    bottomBorderColor: bottomBorderColor,
    bottomBorderStyle: bottomBorderStyle
  }) || Object.keys(additionalProperties).length > 0) {
    return omitUndefinedProperties(_objectSpread(_objectSpread({}, additionalProperties), {}, {
      // alignment
      align: align,
      alignVertical: alignVertical,
      textRotation: textRotation,
      indent: indent,
      wrap: wrap,
      // font
      fontFamily: fontFamily,
      fontSize: fontSize,
      fontWeight: fontWeight,
      fontStyle: fontStyle,
      textDecoration: textDecoration,
      textColor: textColor,
      // fill
      backgroundColor: backgroundColor,
      fillPatternStyle: fillPatternStyle,
      fillPatternColor: fillPatternColor,
      // border
      borderColor: borderColor,
      borderStyle: borderStyle,
      leftBorderColor: leftBorderColor,
      leftBorderStyle: leftBorderStyle,
      rightBorderColor: rightBorderColor,
      rightBorderStyle: rightBorderStyle,
      topBorderColor: topBorderColor,
      topBorderStyle: topBorderStyle,
      bottomBorderColor: bottomBorderColor,
      bottomBorderStyle: bottomBorderStyle
    }));
  }
}
function omitUndefinedProperties(object) {
  var filteredObject = {};
  for (var key in object) {
    if (object[key] !== undefined) {
      filteredObject[key] = object[key];
    }
  }
  return filteredObject;
}
