// write-excel-file/modules/xlsx/initializeSheets.parameters.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import addProperties from './helpers/features/addProperties.js';
export default function getInitializeSheetsParameters(parameters, data, features) {
  var schema = parameters.schema,
    columns = parameters.columns,
    sheetName = parameters.sheet,
    sheetNames = parameters.sheets,
    getHeaderStyle = parameters.getHeaderStyle,
    fontFamily = parameters.fontFamily,
    fontSize = parameters.fontSize,
    orientation = parameters.orientation,
    showGridLines = parameters.showGridLines,
    rightToLeft = parameters.rightToLeft,
    zoomScale = parameters.zoomScale,
    dateFormat = parameters.dateFormat;
  var sheetsParameters = {
    data: data,
    features: features,
    schema: schema,
    columns: columns,
    sheetNames: sheetNames,
    getHeaderStyle: getHeaderStyle,
    fontFamily: fontFamily,
    fontSize: fontSize,
    orientation: orientation,
    showGridLines: showGridLines,
    rightToLeft: rightToLeft,
    zoomScale: zoomScale,
    dateFormat: dateFormat
  };

  // Versions before `1.3.4` had a bug:
  // In a "write multiple sheets" scenario, `columns` parameter
  // wasn't required to be an array of `columns` for each sheet.
  if (sheetNames) {
    if (columns) {
      if (!Array.isArray(columns[0])) {
        throw new Error('In a "write multiple sheets" scenario, `columns` parameter must be an array of `columns` for each sheet.');
      }
    }
  }

  // Add any custom properties that're used by plugins.
  addProperties(sheetsParameters, parameters, features, 'xl/worksheets/sheet{id}.xml');

  // Also add any custom styles properties that're used by plugins
  // because styles properties are read from sheets properties.
  addProperties(sheetsParameters, parameters, features, 'xl/styles.xml');

  // If only a single sheet is being written,
  // convert parameters to arrays as if multiple sheets were being written.
  // This way, the code after this wouldn't bother about the parameters being arrays or not.
  if (!sheetNames) {
    return _objectSpread(_objectSpread({}, sheetsParameters), {}, {
      multipleSheetsParameters: false,
      sheetNames: [sheetName || 'Sheet1']
    });
  }
  return _objectSpread(_objectSpread({}, sheetsParameters), {}, {
    multipleSheetsParameters: true
  });
}
