// write-excel-file/commonjs/xlsx/files/sheet.xml/sheet.xml.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = generateSheetXml;
var _rows = _interopRequireDefault(require("./rows.js"));
var _columns = _interopRequireDefault(require("./columns.js"));
var _processMergedCells2 = _interopRequireDefault(require("./processMergedCells.js"));
var _mergedCellsDescription = _interopRequireDefault(require("./mergedCellsDescription.js"));
var _layout = _interopRequireDefault(require("./layout.js"));
var _views = _interopRequireDefault(require("./views.js"));
var _drawingReference = _interopRequireDefault(require("./drawingReference.js"));
var _getAdditionalContent = _interopRequireDefault(require("../../helpers/features/getAdditionalContent.js"));
var _transformContent = _interopRequireDefault(require("../../helpers/features/transformContent.js"));
var _excluded = ["data", "schema", "columns", "images", "getHeaderStyle", "findOrCreateCellStyle", "findOrCreateSharedString", "customFont", "dateFormat", "orientation", "showGridLines", "rightToLeft", "zoomScale", "sheetIndex", "sheetId", "multipleSheetsParameters"];
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function generateSheetXml(parameters, features) {
  var data_ = parameters.data,
    schema = parameters.schema,
    columns = parameters.columns,
    images = parameters.images,
    getHeaderStyle = parameters.getHeaderStyle,
    findOrCreateCellStyle = parameters.findOrCreateCellStyle,
    findOrCreateSharedString = parameters.findOrCreateSharedString,
    customFont = parameters.customFont,
    dateFormat = parameters.dateFormat,
    orientation = parameters.orientation,
    showGridLines = parameters.showGridLines,
    rightToLeft = parameters.rightToLeft,
    zoomScale = parameters.zoomScale,
    sheetIndex = parameters.sheetIndex,
    sheetId = parameters.sheetId,
    multipleSheetsParameters = parameters.multipleSheetsParameters,
    restParameters = _objectWithoutProperties(parameters, _excluded);
  var _processMergedCells = (0, _processMergedCells2["default"])(data_, {
      schema: schema,
      features: features
    }),
    data = _processMergedCells.data,
    mergedCells = _processMergedCells.mergedCells;
  var xml = '<?xml version="1.0" ?>' + '<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:mv="urn:schemas-microsoft-com:mac:vml" xmlns:mx="http://schemas.microsoft.com/office/mac/excel/2008/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main" xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac" xmlns:xm="http://schemas.microsoft.com/office/excel/2006/main">' +
  // For some weird reason, Excel 2007 demands `<sheetViews/>` element to be the first one.
  (0, _views["default"])({
    showGridLines: showGridLines,
    rightToLeft: rightToLeft,
    zoomScale: zoomScale,
    sheetIndex: sheetIndex
  }) +
  // For some weird reason, Excel 2007 demands `<cols/>` element to follow `<sheetViews/>` element.
  (0, _columns["default"])({
    schema: schema,
    columns: columns
  }) + '<sheetData>' + (0, _rows["default"])(data, {
    schema: schema,
    getHeaderStyle: getHeaderStyle,
    findOrCreateCellStyle: findOrCreateCellStyle,
    findOrCreateSharedString: findOrCreateSharedString,
    customFont: customFont,
    features: features,
    dateFormat: dateFormat
  }) + '</sheetData>' + (0, _mergedCellsDescription["default"])(mergedCells) + (0, _layout["default"])({
    sheetId: sheetId,
    orientation: orientation
  }) + (0, _drawingReference["default"])() +
  // Apply any plugins that insert additional content to this XML.
  (0, _getAdditionalContent["default"])('xl/worksheets/sheet{id}.xml', features, restParameters, {
    multipleSheetsParameters: multipleSheetsParameters,
    sheetIndex: sheetIndex,
    sheetId: sheetId
  }) + '</worksheet>';

  // Apply any plugins that transform this XML.
  xml = (0, _transformContent["default"])(xml, 'xl/worksheets/sheet{id}.xml', features, restParameters, {
    multipleSheetsParameters: multipleSheetsParameters,
    sheetIndex: sheetIndex,
    sheetId: sheetId
  });
  return xml;
}
