// write-excel-file/commonjs/xlsx/files/drawing.xml.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DRAWING_XML_START = exports.DRAWING_XML_END = void 0;
exports["default"] = generateDrawingXml;
var _getAdditionalContent = _interopRequireDefault(require("../helpers/features/getAdditionalContent.js"));
var _transformContent = _interopRequireDefault(require("../helpers/features/transformContent.js"));
var _excluded = ["multipleSheetsParameters", "sheetIndex", "sheetId", "features"];
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
// https://gitlab.com/catamphetamine/write-excel-file/-/blob/main/docs/IMAGES.md
function generateDrawingXml(_ref) {
  var multipleSheetsParameters = _ref.multipleSheetsParameters,
    sheetIndex = _ref.sheetIndex,
    sheetId = _ref.sheetId,
    features = _ref.features,
    restParameters = _objectWithoutProperties(_ref, _excluded);
  var xml = DRAWING_XML_START +
  // Apply any plugins that insert additional content to this XML.
  (0, _getAdditionalContent["default"])('xl/drawings/drawing{id}.xml', features, restParameters, {
    multipleSheetsParameters: multipleSheetsParameters,
    sheetIndex: sheetIndex,
    sheetId: sheetId
  }) + DRAWING_XML_END;

  // Apply any plugins that transform this XML.
  xml = (0, _transformContent["default"])(xml, 'xl/drawings/drawing{id}.xml', features, restParameters, {
    multipleSheetsParameters: multipleSheetsParameters,
    sheetIndex: sheetIndex,
    sheetId: sheetId
  });
  return xml;
}
var DRAWING_XML_START = exports.DRAWING_XML_START = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>' + '<xdr:wsDr xmlns:xdr="http://schemas.openxmlformats.org/drawingml/2006/spreadsheetDrawing" xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main">';
var DRAWING_XML_END = exports.DRAWING_XML_END = '</xdr:wsDr>';
