// write-excel-file/modules/xlsx/files/sheet.xml/sheet.xml.js
var _excluded = ["data", "schema", "columns", "images", "getHeaderStyle", "findOrCreateCellStyle", "findOrCreateSharedString", "customFont", "dateFormat", "orientation", "showGridLines", "rightToLeft", "zoomScale", "sheetIndex", "sheetId", "multipleSheetsParameters"];
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
import generateRows from './rows.js';
import generateColumnsDescription from './columns.js';
import processMergedCells from './processMergedCells.js';
import generateMergedCellsDescription from './mergedCellsDescription.js';
import generateLayout from './layout.js';
import generateViews from './views.js';
import generateDrawingReference from './drawingReference.js';
import getAdditionalContent from '../../helpers/features/getAdditionalContent.js';
import transformContent from '../../helpers/features/transformContent.js';
export default function generateSheetXml(parameters, features) {
  var data_ = parameters.data,
    schema = parameters.schema,
    columns = parameters.columns,
    images = parameters.images,
    getHeaderStyle = parameters.getHeaderStyle,
    findOrCreateCellStyle = parameters.findOrCreateCellStyle,
    findOrCreateSharedString = parameters.findOrCreateSharedString,
    customFont = parameters.customFont,
    dateFormat = parameters.dateFormat,
    orientation = parameters.orientation,
    showGridLines = parameters.showGridLines,
    rightToLeft = parameters.rightToLeft,
    zoomScale = parameters.zoomScale,
    sheetIndex = parameters.sheetIndex,
    sheetId = parameters.sheetId,
    multipleSheetsParameters = parameters.multipleSheetsParameters,
    restParameters = _objectWithoutProperties(parameters, _excluded);
  var _processMergedCells = processMergedCells(data_, {
      schema: schema,
      features: features
    }),
    data = _processMergedCells.data,
    mergedCells = _processMergedCells.mergedCells;
  var xml = '<?xml version="1.0" ?>' + '<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:mv="urn:schemas-microsoft-com:mac:vml" xmlns:mx="http://schemas.microsoft.com/office/mac/excel/2008/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main" xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac" xmlns:xm="http://schemas.microsoft.com/office/excel/2006/main">' +
  // For some weird reason, Excel 2007 demands `<sheetViews/>` element to be the first one.
  generateViews({
    showGridLines: showGridLines,
    rightToLeft: rightToLeft,
    zoomScale: zoomScale,
    sheetIndex: sheetIndex
  }) +
  // For some weird reason, Excel 2007 demands `<cols/>` element to follow `<sheetViews/>` element.
  generateColumnsDescription({
    schema: schema,
    columns: columns
  }) + '<sheetData>' + generateRows(data, {
    schema: schema,
    getHeaderStyle: getHeaderStyle,
    findOrCreateCellStyle: findOrCreateCellStyle,
    findOrCreateSharedString: findOrCreateSharedString,
    customFont: customFont,
    features: features,
    dateFormat: dateFormat
  }) + '</sheetData>' + generateMergedCellsDescription(mergedCells) + generateLayout({
    sheetId: sheetId,
    orientation: orientation
  }) + generateDrawingReference() +
  // Apply any plugins that insert additional content to this XML.
  getAdditionalContent('xl/worksheets/sheet{id}.xml', features, restParameters, {
    multipleSheetsParameters: multipleSheetsParameters,
    sheetIndex: sheetIndex,
    sheetId: sheetId
  }) + '</worksheet>';

  // Apply any plugins that transform this XML.
  xml = transformContent(xml, 'xl/worksheets/sheet{id}.xml', features, restParameters, {
    multipleSheetsParameters: multipleSheetsParameters,
    sheetIndex: sheetIndex,
    sheetId: sheetId
  });
  return xml;
}
