// write-excel-file/commonjs/xlsx/initializeSheets.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = initializeSheets;
var _initializeStyles2 = _interopRequireDefault(require("./initializeStyles.js"));
var _initializeStylesParameters = _interopRequireDefault(require("./initializeStyles.parameters.js"));
var _initializeSharedStrings = _interopRequireDefault(require("./initializeSharedStrings.js"));
var _validateSheetName = _interopRequireDefault(require("./validateSheetName.js"));
var _isObject = _interopRequireDefault(require("./helpers/isObject.js"));
var _excluded = ["sheetNames", "getHeaderStyle", "fontFamily", "fontSize", "features", "multipleSheetsParameters"],
  _excluded2 = ["data", "schema", "columns"];
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _createForOfIteratorHelperLoose(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (t) return (t = t.call(r)).next.bind(t); if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var o = 0; return function () { return o >= r.length ? { done: !0 } : { done: !1, value: r[o++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function initializeSheets(_ref) {
  var sheetNames = _ref.sheetNames,
    getHeaderStyle = _ref.getHeaderStyle,
    fontFamily = _ref.fontFamily,
    fontSize = _ref.fontSize,
    features = _ref.features,
    multipleSheetsParameters = _ref.multipleSheetsParameters,
    restParameters = _objectWithoutProperties(_ref, _excluded);
  var _initializeSharedStri = (0, _initializeSharedStrings["default"])(),
    getSharedStrings = _initializeSharedStri.getSharedStrings,
    findOrCreateSharedString = _initializeSharedStri.findOrCreateSharedString;
  var _initializeStyles = (0, _initializeStyles2["default"])(_objectSpread({
      fontFamily: fontFamily,
      fontSize: fontSize,
      features: features
    }, (0, _initializeStylesParameters["default"])(restParameters, multipleSheetsParameters, features))),
    getCellStyles = _initializeStyles.getCellStyles,
    findOrCreateCellStyle = _initializeStyles.findOrCreateCellStyle;

  // Validate sheet name.
  for (var _iterator = _createForOfIteratorHelperLoose(sheetNames), _step; !(_step = _iterator()).done;) {
    var sheetName = _step.value;
    (0, _validateSheetName["default"])(sheetName);
  }
  var generateSheetXmlParameters = [];
  var sheetIndex = 0;
  for (var _iterator2 = _createForOfIteratorHelperLoose(sheetNames), _step2; !(_step2 = _iterator2()).done;) {
    var _sheetName = _step2.value;
    var data = restParameters.data,
      schema = restParameters.schema,
      columns = restParameters.columns,
      sheetParameters = _objectWithoutProperties(restParameters, _excluded2);
    validateData(data, {
      schema: schema,
      multipleSheetsParameters: multipleSheetsParameters
    });
    // Validate the cases when this function is called directly in tests, etc.
    if (multipleSheetsParameters === undefined) {
      throw new Error('`multipleSheetsParameters` parameter not found');
    }
    var getParameterForSheet = function getParameterForSheet(parameter) {
      if (multipleSheetsParameters) {
        return parameter[sheetIndex];
      }
      return parameter;
    };
    generateSheetXmlParameters.push(_objectSpread(_objectSpread({}, sheetParameters), {}, {
      multipleSheetsParameters: multipleSheetsParameters,
      sheetIndex: sheetIndex,
      sheetId: getSheetId(sheetIndex),
      data: getParameterForSheet(data),
      schema: schema && getParameterForSheet(schema),
      columns: columns && getParameterForSheet(columns),
      customFont: fontFamily || fontSize,
      getHeaderStyle: getHeaderStyle,
      findOrCreateCellStyle: findOrCreateCellStyle,
      findOrCreateSharedString: findOrCreateSharedString
    }));
    sheetIndex++;
  }
  return {
    sheets: sheetNames.map(function (sheetName, _i) {
      // Reassign the `_i` variable to prevent "closure" bug when in `generateSheetXml()` function
      // the `_i` value is always equal to the last sheet's index.
      var sheetIndex = _i;
      return {
        sheetId: getSheetId(sheetIndex),
        sheetName: sheetName,
        generateSheetXmlParameters: generateSheetXmlParameters[sheetIndex]
      };
    }),
    getSharedStrings: getSharedStrings,
    getCellStyles: getCellStyles,
    multipleSheetsParameters: multipleSheetsParameters
  };
}
function getSheetId(sheetIndex) {
  return String(sheetIndex + 1);
}
function validateData(data, _ref2) {
  var schema = _ref2.schema,
    multipleSheetsParameters = _ref2.multipleSheetsParameters;
  if (multipleSheetsParameters) {
    if (!Array.isArray(data) || data.length > 0 && !Array.isArray(data[0])) {
      throw new TypeError('Expected `data` to be an array of arrays of rows');
    }
    for (var _iterator3 = _createForOfIteratorHelperLoose(data), _step3; !(_step3 = _iterator3()).done;) {
      var sheetData = _step3.value;
      validateSheetDataRows(sheetData, {
        schema: schema
      });
    }
  } else {
    if (!Array.isArray(data)) {
      throw new TypeError('Expected `data` to be an array of rows');
    }
    validateSheetDataRows(data, {
      schema: schema
    });
  }
}
function validateSheetDataRows(data, _ref3) {
  var schema = _ref3.schema;
  for (var _iterator4 = _createForOfIteratorHelperLoose(data), _step4; !(_step4 = _iterator4()).done;) {
    var row = _step4.value;
    if (row !== null && row !== undefined) {
      if (schema) {
        if ((0, _isObject["default"])(row)) {
          return;
        } else {
          throw new Error('Expected `data` rows to be objects');
        }
      } else {
        if (Array.isArray(row)) {
          return;
        } else {
          throw new Error('Expected `data` rows to be arrays');
        }
      }
    }
  }
}
