// write-excel-file/modules/xlsx/helpers/features/getAdditionalContent.js
function _createForOfIteratorHelperLoose(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (t) return (t = t.call(r)).next.bind(t); if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var o = 0; return function () { return o >= r.length ? { done: !0 } : { done: !1, value: r[o++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
import attributeValue from '../../../xml/escapeAttributeValue.js';
import textContent from '../../../xml/escapeTextContent.js';

// Returns the content that should be inserted by `features` when performing a `transformName` transform.
export default function getAdditionalContent(transformName, features, availableParameters, _ref) {
  var multipleSheetsParameters = _ref.multipleSheetsParameters,
    sheetIndex = _ref.sheetIndex,
    sheetId = _ref.sheetId;
  var content = '';
  for (var _iterator = _createForOfIteratorHelperLoose(features), _step; !(_step = _iterator()).done;) {
    var feature = _step.value;
    var transform = feature.files && feature.files.transform && feature.files.transform[transformName];
    if (transform && transform.insert) {
      var customParameters = transform.parameters ? transform.parameters(availableParameters) : undefined;
      var insertedContent = transform.insert(customParameters, {
        multipleSheetsParameters: multipleSheetsParameters,
        sheetIndex: sheetIndex,
        sheetId: sheetId,
        attributeValue: attributeValue,
        textContent: textContent
      });
      if (insertedContent) {
        content += insertedContent;
      }
    }
  }
  return content;
}
