// archiver-node/cjs/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collectStream = collectStream;
exports.dateify = dateify;
exports.normalizeInputSource = normalizeInputSource;
exports.sanitizePath = sanitizePath;
exports.trailingSlashIt = trailingSlashIt;
var _normalizePath = _interopRequireDefault(require("normalize-path"));
var _readableStream = require("readable-stream");
var _isStream = require("is-stream");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function collectStream(source, callback) {
  var collection = [];
  var size = 0;
  source.on("error", callback);
  source.on("data", function (chunk) {
    collection.push(chunk);
    size += chunk.length;
  });
  source.on("end", function () {
    var buf = Buffer.alloc(size);
    var offset = 0;
    collection.forEach(function (data) {
      data.copy(buf, offset);
      offset += data.length;
    });
    callback(null, buf);
  });
}
function dateify(dateish) {
  dateish = dateish || new Date();
  if (dateish instanceof Date) {
    dateish = dateish;
  } else if (typeof dateish === "string") {
    dateish = new Date(dateish);
  } else {
    dateish = new Date();
  }
  return dateish;
}
function normalizeInputSource(source) {
  if (source === null) {
    return Buffer.alloc(0);
  } else if (typeof source === "string") {
    return Buffer.from(source);
  } else if ((0, _isStream.isStream)(source, {
    canOpen: false
  })) {
    // Always pipe through a PassThrough stream to guarantee pausing the stream if it's already flowing,
    // since it will only be processed in a (distant) future iteration of the event loop, and will lose
    // data if already flowing now.
    return source.pipe(new _readableStream.PassThrough());
  }
  return source;
}
function sanitizePath(filepath) {
  return (0, _normalizePath["default"])(filepath, false).replace(/^\w+:/, "").replace(/^(\.\.\/|\/)+/, "");
}
function trailingSlashIt(str) {
  return str.slice(-1) !== "/" ? str + "/" : str;
}
