// archiver-node/cjs/plugins/tar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _zlib = _interopRequireDefault(require("zlib"));
var _tarStream = _interopRequireDefault(require("tar-stream"));
var _utils = require("../utils.js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
/**
 * TAR Format Plugin
 *
 * @module plugins/tar
 * @license [MIT]{@link https://github.com/archiverjs/node-archiver/blob/master/LICENSE}
 * @copyright (c) 2012-2014 Chris Talkington, contributors.
 */
var Tar = exports["default"] = /*#__PURE__*/function () {
  /**
   * @constructor
   * @param {TarOptions} options
   */
  function Tar(options) {
    _classCallCheck(this, Tar);
    options = this.options = _objectSpread({
      gzip: false
    }, options);
    if (_typeof(options.gzipOptions) !== "object") {
      options.gzipOptions = {};
    }
    this.engine = _tarStream["default"].pack(options);
    this.compressor = false;
    if (options.gzip) {
      this.compressor = _zlib["default"].createGzip(options.gzipOptions);
      this.compressor.on("error", this._onCompressorError.bind(this));
    }
  }
  /**
   * [_onCompressorError description]
   *
   * @private
   * @param  {Error} err
   * @return void
   */
  return _createClass(Tar, [{
    key: "_onCompressorError",
    value: function _onCompressorError(err) {
      this.engine.emit("error", err);
    }
    /**
     * [append description]
     *
     * @param  {(Buffer|Stream)} source
     * @param  {TarEntryData} data
     * @param  {Function} callback
     * @return void
     */
  }, {
    key: "append",
    value: function append(source, data, callback) {
      var self = this;
      data.mtime = data.date;
      function append(err, sourceBuffer) {
        if (err) {
          callback(err);
          return;
        }
        self.engine.entry(data, sourceBuffer, function (err) {
          callback(err, data);
        });
      }
      if (data.sourceType === "buffer") {
        append(null, source);
      } else if (data.sourceType === "stream" && data.stats) {
        data.size = data.stats.size;
        var entry = self.engine.entry(data, function (err) {
          callback(err, data);
        });
        source.pipe(entry);
      } else if (data.sourceType === "stream") {
        (0, _utils.collectStream)(source, append);
      }
    }
    /**
     * [finalize description]
     *
     * @return void
     */
  }, {
    key: "finalize",
    value: function finalize() {
      this.engine.finalize();
    }
    /**
     * [on description]
     *
     * @return this.engine
     */
  }, {
    key: "on",
    value: function on() {
      return this.engine.on.apply(this.engine, arguments);
    }
    /**
     * [pipe description]
     *
     * @param  {String} destination
     * @param  {Object} options
     * @return this.engine
     */
  }, {
    key: "pipe",
    value: function pipe(destination, options) {
      if (this.compressor) {
        return this.engine.pipe.apply(this.engine, [this.compressor]).pipe(destination, options);
      } else {
        return this.engine.pipe.apply(this.engine, arguments);
      }
    }
    /**
     * [unpipe description]
     *
     * @return this.engine
     */
  }, {
    key: "unpipe",
    value: function unpipe() {
      if (this.compressor) {
        return this.compressor.unpipe.apply(this.compressor, arguments);
      } else {
        return this.engine.unpipe.apply(this.engine, arguments);
      }
    }
  }]);
}();
